### PR TITLE
do not sync IG when not necessary

### PR DESCRIPTION
**Pull request by freehan**

### CODE REVIEW

**Comment by bowei** [comment]
Can you fill out the commit message with details about what is being fixed?
How was this tested?

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -421,6 +399,40 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
...
+	nodeNames, err := utils.GetReadyNodeNames(listers.NewNodeLister(lbc.nodeLister))
+	if err != nil {
+		return err
+	}
+	// Add/remove instances to the instance groups.
+	if err = lbc.instancePool.Sync(nodeNames); err != nil {
+		return err
+	}
+
+	// TODO: Remove this after deprecation
```
Comment (bowei) [inline pkg/controller/controller.go:418]: can we get rid of this yet?

Comment (bowei) [reply pkg/controller/controller.go:418]: @nikhiljindal

**Comment by wojtek-t** [comment]
LGTM

But I will let @bowei to tag it - he is more familiar with the codebase than me.

File: pkg/controller/controller.go (hunk 1 above)
Comment (freehan) [reply pkg/controller/controller.go:418]: I am not sure whether kubemci is still support or not.

**Comment by freehan** [comment]
@bowei Updated the commit message.

**Comment by freehan** [comment]
Added testing. Ready for another round.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -398,6 +373,9 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
 
 	// Get the zones our groups live in.
 	zones, err := lbc.Translator.ListZones()
+	if err != nil {
+		klog.Errorf("Failed to list zones: %v", err)
```
Comment (bowei) [inline pkg/controller/controller.go:377]: Isn't this a problem to go onto the rest of the flow if we aren't able to get a list of the zones?

**Comment by bowei** [comment]
BTW -- can we have commits message that follow this: https://chris.beams.io/posts/git-commit/

For trick change like this, it's worthwhile explaining a bit more about these changes.
The commit messages help the reviewer as well as those who are going to be maintaining the code

**Comment by freehan** [comment]
Updated commit message.
Fix the comments.
Ready for another round.

File: pkg/controller/controller.go (hunk 2 above)
Comment (freehan) [reply pkg/controller/controller.go]: fixed

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -398,6 +373,9 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
 
 	// Get the zones our groups live in.
 	zones, err := lbc.Translator.ListZones()
+	if err != nil {
+		return err
```
Comment (bowei) [inline pkg/controller/controller.go:377]: klog.Errorf("lbc.Translator.ListZones() = %v", err)

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -359,36 +359,11 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
...
-		if ing.Annotations == nil {
-			ing.Annotations = map[string]string{}
-		}
-		if err = setInstanceGroupsAnnotation(ing.Annotations, igs); err != nil {
-			return err
-		}
-		if err = updateAnnotations(lbc.ctx.KubeClient, ing.Name, ing.Namespace, ing.Annotations); err != nil {
+	// Only sync instance group when IG is used for this ingress
+	if len(nodePorts(ingSvcPorts)) > 0 {
+		if err := lbc.syncInstanceGroup(syncState.ing, ingSvcPorts); err != nil {
```
Comment (bowei) [inline pkg/controller/controller.go:364]: log error

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -421,6 +399,40 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
 	return nil
 }
 
+// syncInstanceGroup creates instance groups, syncs instances, sets named ports and updates instance group annotation
+func (lbc *LoadBalancerController) syncInstanceGroup(ing *v1beta1.Ingress, ingSvcPorts []utils.ServicePort) error {
+	igs, err := lbc.instancePool.EnsureInstanceGroupsAndPorts(lbc.ctx.ClusterNamer.InstanceGroup(), nodePorts(ingSvcPorts))
+	if err != nil {
+		return err
```
Comment (bowei) [inline pkg/controller/controller.go:406]: you need to log errors or at the level above

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -421,6 +399,40 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
 	return nil
 }
 
+// syncInstanceGroup creates instance groups, syncs instances, sets named ports and updates instance group annotation
+func (lbc *LoadBalancerController) syncInstanceGroup(ing *v1beta1.Ingress, ingSvcPorts []utils.ServicePort) error {
+	igs, err := lbc.instancePool.EnsureInstanceGroupsAndPorts(lbc.ctx.ClusterNamer.InstanceGroup(), nodePorts(ingSvcPorts))
```
Comment (bowei) [inline pkg/controller/controller.go:404]: log the action taking place here

klog.V(2).Infof(

ns, name of ingress
svc ports,

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -421,6 +399,40 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
...
+		return err
+	}
+	// Add/remove instances to the instance groups.
+	if err = lbc.instancePool.Sync(nodeNames); err != nil {
+		return err
+	}
+
+	// TODO: Remove this after deprecation
+	if utils.IsGCEMultiClusterIngress(ing) {
+		// Add instance group names as annotation on the ingress and return.
```
Comment (bowei) [inline pkg/controller/controller.go:420]: log a warning here that kubemci is still enabled

File: pkg/controller/controller.go [hunk 8]
```diff
@@ -359,36 +359,11 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
...
-	if utils.IsGCEMultiClusterIngress(syncState.ing) {
-		// Add instance group names as annotation on the ingress and return.
-		if ing.Annotations == nil {
-			ing.Annotations = map[string]string{}
-		}
-		if err = setInstanceGroupsAnnotation(ing.Annotations, igs); err != nil {
-			return err
-		}
-		if err = updateAnnotations(lbc.ctx.KubeClient, ing.Name, ing.Namespace, ing.Annotations); err != nil {
+	// Only sync instance group when IG is used for this ingress
```
Comment (bowei) [inline pkg/controller/controller.go:362]: else {
  // log skipping IG sync
}

**Comment by bowei** [comment]
need to log major actions

**Comment by freehan** [comment]
Fix the comments.
Ready for another round.

File: pkg/controller/controller.go [hunk 9]
```diff
@@ -359,36 +359,14 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
...
+	if len(nodePorts(ingSvcPorts)) > 0 {
+		if err := lbc.syncInstanceGroup(syncState.ing, ingSvcPorts); err != nil {
+			klog.Errorf("Failed to sync instance group: %v", err)
 			return err
 		}
-		// This short-circuit will stop the syncer from moving to next step.
-		return ingsync.ErrSkipBackendsSync
+	} else {
+		klog.V(2).Infof("Skip syncing instance groups")
```
Comment (bowei) [inline pkg/controller/controller.go:369]: log the identity of the ingress object that this is considering 

syncState.ing

File: pkg/controller/controller.go (hunk 4 above)
Comment (bowei) [reply pkg/controller/controller.go:364]: log which ingress this was looking at syncState.ing.

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -421,6 +403,43 @@ func (lbc *LoadBalancerController) SyncBackends(state interface{}) error {
...
+		return err
+	}
+	// Add/remove instances to the instance groups.
+	if err = lbc.instancePool.Sync(nodeNames); err != nil {
+		return err
+	}
+
+	// TODO: Remove this after deprecation
+	if utils.IsGCEMultiClusterIngress(ing) {
+		klog.Warningf("kubemci is used for ingress %v+", ing)
```
Comment (bowei) [inline pkg/controller/controller.go:426]: What is with the +

File: pkg/controller/controller_test.go [hunk 11]
```diff
@@ -176,6 +178,37 @@ func TestIngressSyncError(t *testing.T) {
 	}
 }
 
+// TestNegOnlyIngress asserts that `sync` will not create IG when there is only NEG backends for the ingress
+func TestNegOnlyIngress(t *testing.T) {
```
Comment (bowei) [inline pkg/controller/controller_test.go:182]: TestNEGOnlyIngress

File: pkg/controller/controller_test.go [hunk 12]
```diff
@@ -176,6 +178,37 @@ func TestIngressSyncError(t *testing.T) {
...
+	lbc := newLoadBalancerController()
+
+	svc := test.NewService(types.NamespacedName{Name: "my-service", Namespace: "default"}, api_v1.ServiceSpec{
+		Type:  api_v1.ServiceTypeNodePort,
+		Ports: []api_v1.ServicePort{{Port: 80}},
+	})
+	negAnnotation := annotations.NegAnnotation{Ingress: true}
+	svc.Annotations = map[string]string{}
```
Comment (bowei) [inline pkg/controller/controller_test.go:190]: combine

```
svc.Annotations = map[string]string{
  annotations.NEGAnnotationKey: annotations.NegAnnotation{Ingress: true}.String(),
}

```

File: pkg/controller/controller_test.go [hunk 13]
```diff
@@ -176,6 +178,37 @@ func TestIngressSyncError(t *testing.T) {
...
+
+	ingStoreKey := getKey(ing, t)
+	err := lbc.sync(ingStoreKey)
+	if err != nil {
+		t.Fatalf("lbc.sync(%v) = %v, want nil", ingStoreKey, err)
+	}
+
+	ig, err := lbc.instancePool.Get(lbc.ctx.ClusterNamer.InstanceGroup(), fakeZone)
+	if err != nil && !utils.IsHTTPErrorCode(err, http.StatusNotFound) {
+		t.Errorf("lbc.ctx.Cloud.ListInstanceGroups(%v) = %v, %v, want [], nil", fakeZone, ig, err)
```
Comment (bowei) [inline pkg/controller/controller_test.go:208]: want [], http.StatusNotFound

**Comment by freehan** [comment]
Fixed. Ready.

File: pkg/controller/controller.go (hunk 4 above)
Comment (bowei) [reply pkg/controller/controller.go:364]: fix this later, you should always use the right type vs %v (%s:%s)

**Pull request merged**